Mathlib/Data/Matroid/Dual.lean

```diff
@@ -0,0 +1,264 @@
+  rw [base_compl_iff_mem_maximals_disjoint_base, base_iff_maximal_indep, dual_indep_iff_exists',
+    mem_maximals_setOf_iff]
+  simp [dual_indep_iff_exists']
+
+theorem dual_base_iff' : M﹡.Base B ↔ M.Base (M.E \ B) ∧ B ⊆ M.E :=
+  (em (B ⊆ M.E)).elim (fun h ↦ by rw [dual_base_iff, and_iff_left h])
+    (fun h ↦ iff_of_false (h ∘ (fun h' ↦ h'.subset_ground)) (h ∘ And.right))
+
+theorem setOf_dual_base_eq : setOf M﹡.Base = (fun X ↦ M.E \ X) '' {B | M.Base B} := by
+  ext B
+  simp only [mem_setOf_eq, mem_image, dual_base_iff']
+  refine' ⟨fun h ↦ ⟨_, h.1, diff_diff_cancel_left h.2⟩,
+    fun ⟨B', hB', h⟩ ↦ ⟨_,h.symm.trans_subset (diff_subset _ _)⟩⟩
+  rwa [←h, diff_diff_cancel_left hB'.subset_ground]
+
+@[simp] theorem dual_dual (M : Matroid α) : M﹡﹡ = M :=
+  eq_of_base_iff_base_forall rfl (fun B (h : B ⊆ M.E) ↦
+    by rw [dual_base_iff, dual_base_iff, dual_ground, diff_diff_cancel_left h])
+
+theorem Base.compl_base_of_dual (h : M﹡.Base B) : M.Base (M.E \ B) :=
+  (dual_base_iff'.1 h).1
+
+theorem Base.compl_base_dual (h : M.Base B) : M﹡.Base (M.E \ B) := by
+  rwa [dual_base_iff, diff_diff_cancel_left h.subset_ground]
+
+theorem Base.compl_inter_basis_of_inter_basis (hB : M.Base B) (hBX : M.Basis (B ∩ X) X) :
+    M﹡.Basis ((M.E \ B) ∩ (M.E \ X)) (M.E \ X) := by
+  refine' Indep.basis_of_forall_insert _ (inter_subset_right _ _) (fun e he ↦ _)
+  · rw [dual_indep_iff_exists]
+    exact ⟨B, hB, disjoint_of_subset_left (inter_subset_left _ _) disjoint_sdiff_left⟩
+  simp only [diff_inter_self_eq_diff, mem_diff, not_and, not_not, imp_iff_right he.1.1] at he
+  simp_rw [dual_dep_iff_forall, insert_subset_iff, and_iff_right he.1.1,
+    and_iff_left ((inter_subset_left _ _).trans (diff_subset _ _))]
+  refine' fun B' hB' ↦ by_contra (fun hem ↦ _)
+  rw [nonempty_iff_ne_empty, not_ne_iff, ←union_singleton, diff_inter_diff,
+   inter_distrib_right, union_empty_iff, singleton_inter_eq_empty, diff_eq,
+   inter_right_comm, inter_eq_self_of_subset_right hB'.subset_ground, ←diff_eq,
+   diff_eq_empty] at hem
+  obtain ⟨f, hfb, hBf⟩ := hB.exchange hB' ⟨he.2, hem.2⟩
+
+  have hi : M.Indep (insert f (B ∩ X))
+  · refine' hBf.indep.subset (insert_subset_insert _)
+    simp_rw [subset_diff, and_iff_right (inter_subset_left _ _), disjoint_singleton_right,
+      mem_inter_iff, iff_false_intro he.1.2, and_false, not_false_iff]
+  exact hfb.2 (hBX.mem_of_insert_indep (Or.elim (hem.1 hfb.1) (False.elim ∘ hfb.2) id) hi).1
+
+theorem Base.inter_basis_iff_compl_inter_basis_dual (hB : M.Base B) (hX : X ⊆ M.E := by aesop_mat):
+    M.Basis (B ∩ X) X ↔ M﹡.Basis ((M.E \ B) ∩ (M.E \ X)) (M.E \ X) := by
+  refine' ⟨hB.compl_inter_basis_of_inter_basis, fun h ↦ _⟩
+  simpa [inter_eq_self_of_subset_right hX, inter_eq_self_of_subset_right hB.subset_ground] using
+    hB.compl_base_dual.compl_inter_basis_of_inter_basis h
+
+theorem dual_inj {M₁ M₂ : Matroid α} (h : M₁﹡ = M₂﹡) : M₁ = M₂ :=
+by rw [←dual_dual M₁, h, dual_dual]
```

Comment:
```suggestion
theorem dual_inj {M₁ M₂ : Matroid α} (h : M₁﹡ = M₂﹡) : M₁ = M₂ := by
  rw [←dual_dual M₁, h, dual_dual]
```
And could you move this theorem to after `dual_dual`? The two after that should move too.